### ISSUE TITLE
Strange behaviour with queries using recursive functions

### ISSUE THREAD

**Issue by niieani** [rethinkdb/rethinkdb#4024]
Hello,

Since RethinkDB doesn't support recursive queries natively yet (https://github.com/rethinkdb/rethinkdb/issues/2751), I've developed this workaround:

```javascript
var howManyRecursions = 3;

var recursiveJoin = function(bubble, count) {
  if (count <= 0)
    return bubble;
  return r.branch(
    // IF
    bubble.hasFields('components'),
    // THEN
    bubble.merge(
      r.object(
        'components', bubble.getField('components')
        .eqJoin(function(t){return t}, r.table("bubbles"))
        .zip().map(function(innerBubble){ return recursiveJoin(innerBubble, count-1) })
      )
    ),
    // ELSE
    bubble
  )
};

r.table('bubbles')
  .get('79b139b3-789a-f39a-5406-b5f2a563db0f').merge(
    r.object(
      'components', r.row('components')
      .eqJoin(function(t){return t}, r.table("bubbles"))
      .zip().map(function(innerBubble){ return recursiveJoin(innerBubble, howManyRecursions) })
    ));
```

This works rather well with the official JS client as long as the number of recursions is small (something like 15 recursions takes 50ms on my computer), but when I do 150, RethinkDb's process stops responding or receiving connections again and needs to be restarted.

If I understand it correctly, in the above case the function is not reused by the server, but actually repeated by the client as many times as the recursion happens (calculated locally).

So I've tried reimplementing the same concept using server-side concepts, like this:
```javascript
var recurringFunction = function(count) {
  return r.branch(
    // IF
    r.expr(count).gt(4),
    // THEN
    recurringFunction(r.expr(count).sub(1)),
    // ELSE
    r.expr(count)
  );
};

r.json('[1,2,3]').map(function(item){
  return recurringFunction(6);
})
```
But I get an error: 
```
JavaScript error:
RangeError: Maximum call stack size exceeded.
```

Any ideas why this is happening?

If you want to test the first implementation, create here's a sample table (```bubbles```) with level 2 join recursion:

```json
[
  {
    "components": [
      "44ecf62e-801d-4215-b5d1-f9e229865954",
      "fcf22d57-f86f-dbe3-5638-d60fe6a7aef7",
      "31cba1e8-bcda-c755-288c-271c2f9afa67"
    ],
    "id": "79b139b3-789a-f39a-5406-b5f2a563db0f"
  },
  {
    "id": "cada4674-cc8a-457e-ba99-b3aefb417fb5",
    "value": "inner1"
  },
  {
    "components": [
      "cada4674-cc8a-457e-ba99-b3aefb417fb5",
      "7c482d67-31b6-26f5-0f3b-6ffc7704697f"
    ],
    "id": "31cba1e8-bcda-c755-288c-271c2f9afa67"
  },
  {
    "id": "44ecf62e-801d-4215-b5d1-f9e229865954",
    "value": "test2"
  },
  {
    "id": "fcf22d57-f86f-dbe3-5638-d60fe6a7aef7",
    "value": "test1"
  },
  {
    "id": "7c482d67-31b6-26f5-0f3b-6ffc7704697f",
    "value": "inner2"
  }
]
```

**Comment by mlucy** [comment]
Hi @niieani ,

In the first case you're generating a very deeply nested query that's probably breaking something on the server (maybe https://github.com/rethinkdb/rethinkdb/issues/2639 ?).  We should detect those queries and reject them, probably.  (Putting this into 1.16.x since hanging forever on a query seems like a bug to me even if it's really deeply nested like that.)

In the second case you're actually recursing forever in the JS client; the `r.branch` gets run on the server, but to construct it the JS client is recursing right away to try and generate one of the branches.  We've talked about adding first-class functions to the server, which would make this possible, but it isn't on the roadmap right now.

Unfortunately I don't think there's anything to do on this except to keep the number of recursions small for now.

**Comment by niieani** [comment]
Hi @mlucy,

Thanks for getting back to me so quickly. Would you have any suggestions on how I could achieve the same effect in any other, better way? I've just looked at the docs, there's the ```r.js```, which, as I understand, seems to be computed on the server. Is it possible it could help me in any way to construct a recursive function on the server, rather than on the client?

**Comment by mlucy** [comment]
Unfortunately not; `r.js` can't call back into ReQL code to do joins.  The server just doesn't support this behavior at the moment, so generating nested queries on the client is about the best you can do.  (We also don't really optimize for this case, so I wouldn't be shocked if there were performance problems for deeply nested joins.)

You could also consider not doing the join on the server, and instead just reading the values in on the client and issue new read queries as needed.

**Comment by niieani** [comment]
OK, I understand. Well, I hope recursive queries do get some love and get to be on the roadmap, I think the there are plenty of great use cases for them.

A good and simple example is a situation where we want to create a tree of one-to-[many of the same] relationships, like, say, a Social Graph of Friends - 1st degree, 2nd degree, 3rd degree connections, etc. built with each recursion.

Going back and forth with the joins and new read queries can work with a small amount of data, but if there'll be thousands of records, it would get very slow with all the round-trips.

Thanks, @mlucy!

**Comment by mlucy** [comment]
@niieani -- another option for doing it in the client is to locate a helper client on the same machine as the DB and have it handle the recursion.  That way there isn't much network latency, and it should be closer to what we could get with native ReQL recursive queries.

**Comment by danielmewes** [comment]
I can reproduce the crash with 15 recursions already:
```
Version: rethinkdb 2.0.0+0RC1~0precise (GCC 4.6.3)
error: Error in src/arch/runtime/thread_pool.cc at line 341:
error: Callstack overflow in a coroutine
error: Backtrace:
addr2line: 'rethinkdb': No such file
addr2line: 'rethinkdb': No such file
error: Tue Apr  7 16:53:41 2015
       
       1: format_backtrace(bool) at 0xb21ae3 (rethinkdb)
       2: report_fatal_error(char const*, int, char const*, ...) at 0xd2ef54 (rethinkdb)
       3: linux_thread_pool_t::sigsegv_handler(int, siginfo*, void*) at 0xa35fb4 (rethinkdb)
       4: /lib/x86_64-linux-gnu/libpthread.so.0(+0xfcb0) [0x7ff0a7548cb0] at 0x7ff0a7548cb0 (/lib/x86_64-linux-gnu/libpthread.so.0)
       5: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x980513 (rethinkdb)
       6: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       7: ql::make_get_all_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x905ae3 (rethinkdb)
       8: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981719 (rethinkdb)
       9: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       10: counted_t<ql::default_term_t> make_counted<ql::default_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8f2c6b (rethinkdb)
       11: ql::make_default_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8f285c (rethinkdb)
       12: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981425 (rethinkdb)
       13: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       14: counted_t<ql::map_term_t> make_counted<ql::map_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8e6701 (rethinkdb)
       15: ql::make_map_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8e19dc (rethinkdb)
       16: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981ee6 (rethinkdb)
       17: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       18: counted_t<ql::branch_term_t> make_counted<ql::branch_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8fbf2b (rethinkdb)
       19: ql::make_branch_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8fa5cc (rethinkdb)
       20: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x9819d6 (rethinkdb)
       21: ql::func_term_t::func_term_t(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8cdfb3 (rethinkdb)
       22: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981914 (rethinkdb)
       23: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       24: ql::make_concatmap_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8e170c (rethinkdb)
       25: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981e7a (rethinkdb)
       26: rethinkdb() [0x8f3857] at 0x8f3857 ()
       27: counted_t<ql::eq_join_term_t> make_counted<ql::eq_join_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8f4a87 (rethinkdb)
       28: ql::make_eq_join_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8f3a9c (rethinkdb)
       29: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981cca (rethinkdb)
       30: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       31: counted_t<ql::zip_term_t> make_counted<ql::zip_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8e6aeb (rethinkdb)
       32: ql::make_zip_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8e1a6c (rethinkdb)
       33: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x98185d (rethinkdb)
       34: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       35: counted_t<ql::map_term_t> make_counted<ql::map_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8e6701 (rethinkdb)
       36: ql::make_map_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8e19dc (rethinkdb)
       37: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981ee6 (rethinkdb)
       38: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       39: counted_t<ql::object_term_t> make_counted<ql::object_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8ed63d (rethinkdb)
       40: ql::make_object_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8ec57c (rethinkdb)
       41: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x980be4 (rethinkdb)
       42: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       43: rethinkdb() [0x92663f] at 0x92663f ()
       44: counted_t<ql::merge_term_t> make_counted<ql::merge_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x928eed (rethinkdb)
       45: ql::make_merge_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x92687c (rethinkdb)
       46: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981f88 (rethinkdb)
       47: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       48: counted_t<ql::branch_term_t> make_counted<ql::branch_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8fbf2b (rethinkdb)
       49: ql::make_branch_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8fa5cc (rethinkdb)
       50: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x9819d6 (rethinkdb)
       51: ql::func_term_t::func_term_t(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8cdfb3 (rethinkdb)
       52: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981914 (rethinkdb)
       53: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       54: counted_t<ql::map_term_t> make_counted<ql::map_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8e6701 (rethinkdb)
       55: ql::make_map_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8e19dc (rethinkdb)
       56: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981ee6 (rethinkdb)
       57: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       58: counted_t<ql::object_term_t> make_counted<ql::object_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8ed63d (rethinkdb)
       59: ql::make_object_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8ec57c (rethinkdb)
       60: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x980be4 (rethinkdb)
       61: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       62: rethinkdb() [0x92663f] at 0x92663f ()
       63: counted_t<ql::merge_term_t> make_counted<ql::merge_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x928eed (rethinkdb)
       64: ql::make_merge_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x92687c (rethinkdb)
       65: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981f88 (rethinkdb)
       66: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       67: counted_t<ql::branch_term_t> make_counted<ql::branch_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8fbf2b (rethinkdb)
       68: ql::make_branch_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8fa5cc (rethinkdb)
       69: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x9819d6 (rethinkdb)
       70: ql::func_term_t::func_term_t(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8cdfb3 (rethinkdb)
       71: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981914 (rethinkdb)
       72: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       73: counted_t<ql::map_term_t> make_counted<ql::map_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8e6701 (rethinkdb)
       74: ql::make_map_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8e19dc (rethinkdb)
       75: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981ee6 (rethinkdb)
       76: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       77: counted_t<ql::object_term_t> make_counted<ql::object_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8ed63d (rethinkdb)
       78: ql::make_object_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8ec57c (rethinkdb)
       79: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x980be4 (rethinkdb)
       80: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       81: rethinkdb() [0x92663f] at 0x92663f ()
       82: counted_t<ql::merge_term_t> make_counted<ql::merge_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x928eed (rethinkdb)
       83: ql::make_merge_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x92687c (rethinkdb)
       84: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981f88 (rethinkdb)
       85: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       86: counted_t<ql::branch_term_t> make_counted<ql::branch_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8fbf2b (rethinkdb)
       87: ql::make_branch_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8fa5cc (rethinkdb)
       88: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x9819d6 (rethinkdb)
       89: ql::func_term_t::func_term_t(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8cdfb3 (rethinkdb)
       90: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981914 (rethinkdb)
       91: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       92: counted_t<ql::map_term_t> make_counted<ql::map_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8e6701 (rethinkdb)
       93: ql::make_map_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8e19dc (rethinkdb)
       94: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x981ee6 (rethinkdb)
       95: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       96: counted_t<ql::object_term_t> make_counted<ql::object_term_t, ql::compile_env_t*&, ql::protob_t<Term const> const&>(ql::compile_env_t*&&&, ql::protob_t<Term const> const&&&) at 0x8ed63d (rethinkdb)
       97: ql::make_object_term(ql::compile_env_t*, ql::protob_t<Term const> const&) at 0x8ec57c (rethinkdb)
       98: ql::compile_term(ql::compile_env_t*, ql::protob_t<Term const>) at 0x980be4 (rethinkdb)
       99: ql::op_term_t::op_term_t(ql::compile_env_t*, ql::protob_t<Term const>, ql::argspec_t, ql::optargspec_t) at 0x89275f (rethinkdb)
       100: rethinkdb() [0x92663f] at 0x92663f ()
```

**Comment by danielmewes** [comment]
We should probably have a nesting limit for term evaluation.

**Comment by mlucy** [comment]
@danielmewes -- we've talked about this before, but I can't remember the reason why we don't just have a guard page and grow coroutine stacks when they get too large.

**Comment by mlucy** [comment]
Wait, nevermind, I remembered.

**Comment by danielmewes** [comment]
We do tell the operating system to unpage most of the coroutine stack now, so we might get away with just making the stacks really big. I'd still be cautious with this though, since we don't have a guarantee that the OS actually obeys our hints and we don't know if it does so immediately or with a delay etc.

If we don't want to impose a hard limit on recursion depth, we could spawn a new fresh coroutine after a certain call depth in the QL code and continue evaluating in that coroutine.

**Comment by danielmewes** [comment]
Talked with @mlucy about this.

We think that we should do the following:
Every time `eval` is called on a term, we should check that we have at least a certain amount of stack space available on the current coroutine (say 4 KB).
I'm not sure if our coroutines already have a method for getting that information, but it should be easy to add. The underlying implementation definitely has functions for getting the stack boundaries.
If we don't have enough stack space available, we can throw an exception and stop the query.

Assigning @Tryneus .

**Comment by danielmewes** [comment]
Working on this

**Comment by danielmewes** [comment]
A fix is up in CR 2955 by @mlucy

**Comment by danielmewes** [comment]
Fixed in v2.0.x as of https://github.com/rethinkdb/rethinkdb/commit/195f365c4e9473de5bfeb2d3bcc8dde488c37c55 and cherry-picked into next in https://github.com/rethinkdb/rethinkdb/commit/4ad7aa042c0bafe247e88d5e5324ad4b1b4d8e90 and https://github.com/rethinkdb/rethinkdb/commit/9ca2bacb43ed4464e02db65ea66d4fc02aab7afa

We now check that we have at least 4 KB of stack space available before compiling or evaluating another term, and fail with a ReQL exception if that's not the case.

**Comment by danielmewes** [comment]
Marking a duplicate of https://github.com/rethinkdb/rethinkdb/issues/2639
